kaiax/valset/impl/schema.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2024 The Kaia Authors
+// This file is part of the Kaia library.
+//
+// The Kaia library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Kaia library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Kaia library. If not, see <http://www.gnu.org/licenses/>.
+
+package impl
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"slices"
+	"sync"
+
+	"github.com/kaiachain/kaia/common"
+	"github.com/kaiachain/kaia/storage/database"
+)
+
+var (
+	validatorVoteBlockNums           = []byte("validatorVoteBlockNums")
+	lowestScannedValidatorVoteNumKey = []byte("lowestScannedValidatorVoteNum")
+	councilPrefix                    = []byte("council")
+	istanbulSnapshotKeyPrefix        = []byte("snapshot")
+
+	mu = &sync.RWMutex{}
+
+	validatorVoteBlockNumsCache []uint64
+)
+
+func councilKey(num uint64) []byte {
+	return append(councilPrefix, common.Int64ToByteLittleEndian(num)...)
+}
+
+func istanbulSnapshotKey(hash common.Hash) []byte {
+	return append(istanbulSnapshotKeyPrefix, hash[:]...)
+}
+
+func ReadValidatorVoteBlockNums(db database.Database) []uint64 {
```

Comment:
I thought the number of historic votes is not small, but do we have expected numbers?

Reply:
That is why I introduced the cache.

There were 210,000 votes until 2023.03, but most of them are ineffective votes mistakenly casted. There are duplicate votes (e.g. trying to add already existing address). I suspect actual number of council changes are much smaller. Also consider that this chain is around 2000 days old and the council change are manual commands.

Reply:
Yes, thanks for confirming.

If so, I think we can consider:
- Append `nums` into cache, not purge them since it's append-only.
- Use `slices.BinarySearch(nums, num)` in `insertValidatorVoteBlockNums` since they're sorted.

Reply:
Allow me to optimize ValidatorVoteBlockNums in another PR. This PR is blocking #183.
My plan is to make an write-through cache layer `NumsCache` where it centrally handles Read, Write, Insert, Trim operations with an in-memory array.

Reply:
Ok, thanks for letting me know.